### PR TITLE
Report page

### PR DESCRIPTION
**Pull request by t-kanin**

### CODE REVIEW

File: app/models/keyword.rb [hunk 1]
```diff
@@ -2,15 +2,18 @@
...
-  has_many :search_result
-  has_many :ad_result
+  has_many :search_result, dependent: :destroy
+  has_many :ad_result, dependent: :destroy
 
   def self.import(file, uid)
     options = {
-      required_headers: [:keyword]
+      required_headers: [:keyword],
+      chunk_size: 10002
```
Comment (aunsira) [inline app/models/keyword.rb]: Please avoid magic number. Use a constant instead.

File: app/models/keyword.rb [hunk 2]
```diff
@@ -2,15 +2,18 @@
...
     }
     SmarterCSV.process(file.path, options) do |chunk|
       Keyword.where(user_id: uid).insert_all(chunk)
+      key_arr = chunk.map { |k| k[:keyword] }
+      ScraperWorker.perform_async(key_arr)
```
Comment (aunsira) [inline app/models/keyword.rb]:
```ruby
ScraperWorker.perform_async(chunk.map(&:keyword))
```

File: app/services/scraper_service.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    @dic.at_css('div#result-stats').text
+  end
+
+  def update_database(key, res)
+    return if res[:page_result].nil?
+
+    Keyword.where(keyword: key).update(page_result: res[:page_result])
+    kid = Keyword.where(keyword: key).last.id
+    SearchResult.where(keyword_id: kid).insert_all(res[:search_result]) unless res[:search_result].empty?
```
Comment (aunsira) [inline app/services/scraper_service.rb]: We could directly query the assoc without id.
```suggestion
    SearchResult.where(keyword: keyword).insert_all(res[:search_result]) unless res[:search_result].empty?
```

File: app/models/keyword.rb [hunk 4]
```diff
@@ -2,15 +2,18 @@
 
 class Keyword < ApplicationRecord
   belongs_to :user
-  has_many :search_result
-  has_many :ad_result
+  has_many :search_result, dependent: :destroy
+  has_many :ad_result, dependent: :destroy
```
Comment (aunsira) [inline app/models/keyword.rb]: Pluralize assoc.

File: app/services/scraper_service.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        .reject { |h| h[:title].empty? }
+  end
+
+  def page_result
+    return 'no results containing your search terms' if @dic.at_css('div#result-stats').nil?
+
+    @dic.at_css('div#result-stats').text
+  end
+
+  def update_database(key, res)
```
Comment (aunsira) [inline app/services/scraper_service.rb]: I think we can have a better method name than update_database. What about store_result?

File: app/services/scraper_service.rb (hunk 3 above)
Comment (aunsira) [reply app/services/scraper_service.rb]:
```suggestion
    SearchResult.where(keyword_id: kid).insert_all(res[:search_result]) if res[:search_result].present?
```

Comment (aunsira) [reply app/services/scraper_service.rb]: Actually do we need `.where(keyword_id: kid)` before inserting?

File: config/routes.rb [hunk 6]
```diff
@@ -6,10 +6,11 @@
...
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
   mount Sidekiq::Web => '/sidekiq'
-  
+
   resources :keywords do
     collection { post :import }
   end
-  
+
+  get 'keywords/:id/:status', to: 'keywords#show'
```
Comment (aunsira) [inline config/routes.rb:14]: Don't think we need this.

File: db/migrate/20210920070835_add_time_stamp.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class AddTimeStamp < ActiveRecord::Migration[6.1]
+  def change
+    change_column_default :search_results, :created_at, -> { 'CURRENT_TIMESTAMP' }
+    change_column_default :search_results, :updated_at, -> { 'CURRENT_TIMESTAMP' }
+    change_column_default :ad_results, :created_at, -> { 'CURRENT_TIMESTAMP' }
+    change_column_default :ad_results, :updated_at, -> { 'CURRENT_TIMESTAMP' }
```
Comment (aunsira) [inline db/migrate/20210920070835_add_time_stamp.rb:6]: We normally don't add default value to `created_at`/`updated_at`. Rails already does this.

File: app/controllers/keywords_controller.rb [hunk 8]
```diff
@@ -24,4 +30,16 @@ def import
...
+  def find_keyword
+    @keyword = Keyword.find(params[:id])
+  end
+
+  def find_links
+    @links = if params[:status].eql?('ads')
+               @keyword.ad_result
+             else
+               @keyword.search_result
+             end
```
Comment (aunsira) [inline app/controllers/keywords_controller.rb]: Consider using switch-case.

File: db/migrate/20210920070835_add_time_stamp.rb (hunk 7 above)
Comment (t-kanin) [reply db/migrate/20210920070835_add_time_stamp.rb:6]: https://github.com/rails/rails/issues/35493 getting this issue when using insert_all kub. maybe better way to fix it ?

Comment (aunsira) [reply db/migrate/20210920070835_add_time_stamp.rb:6]: Then we should prepare the data up first.

```ruby
bulk_data = {
  ...
  created_at: Time.current,
  updated_at: Time.current
}
Foo.insert_all(bulk_data)

```

**Review by aunsira — APPROVED** [review]

**Pull request merged**